README.md
Instruction on Examcopedia
======

The main app is a `node` app with a `mysql` database.

You can checkout the repository here:

```
https://github.com/xgui3783/examcopedia
```

You will want to use the branch `without_email_confirmation`. 

If you want to set it up manually, you will need to do the following:

- set up VM, install & run mysql, install node (> v6)
- setup the following environmental variables: DB_HOST, MYSQL_DB_USERNAME, MYSQL_DB_PASSWORD
- in the root directory, type `node server.js`

Alternatively, if you have docker and docker-compose configured, you can download/save the following script:

```
version: '2'

services:
  app:
    image: xgui3783/examcopedia:no_email_verification
    depends_on:
      - mysql
    volumes :
      - ${PWD}/config.js:/examcopedia/public/include/config.js
      - ${PWD}/data/img:/examcopedia/public/img
    ports:
      - 80:3002
    environment:
      - DB_HOST=mysql
      - MYSQL_DB_USERNAME=root
      - MYSQL_DB_PASSWORD=SOMETHINGRANDOMPASSWORD_IS_A_JELLYFISH_
  mysql:
    image: mysql:5
    environment:
      - MYSQL_ROOT_PASSWORD=SOMETHINGRANDOMPASSWORD_IS_A_JELLYFISH_
  phpmyadmin:
    image: phpmyadmin/phpmyadmin
    depends_on:
      - mysql
    ports:
      - 127.0.0.1:81:80
    environment:
      - MYSQL_ROOT_PASSWORD=SOMETHINGRANDOMPASSWORD_IS_A_JELLYFISH_
      - PMA_HOST=mysql

```

then type:

`docker-compose up`

and it should work out of the box. I may have to make some adjustments to `xgui3783/examcopedia:no_email_verification`, that will happen next weekend. 


changelog.txt
Known issues:
- filename unescaped
- uploading image with same filename will overwrite the original
- potential vulnerability via uploaded image

Realised goals:
- implement cold-edit of questions
- implement login/register/signin via google/FB/etc, moderator status
- generate printable PDF
- implement pasting of image from clipboard
- API allowing others to interact with the webapp
- answers to be printed in pdf
- implement moderator approval process of addition/edition of questions
- local registration

Currently worked on goals:
- username/password/account locked status message in login screen

Backburner:
- implement "swap out" button in view/generate > random, so that educators can swap out the questions they don't like and replace it with a new random question

Future updates:

Change log
v0.1.11 - 24/10/2016 - categorise mode now allows SHOW ONLY or EXCLUDE mode for syllabus (previously, only EXCLUDE was implemented)
			- added a new API, allows users to call /categoriseQ, displaying the categorisation and description of the categories, given the hash ids
			- improved PDF
			- added functionality for custom PDF (consult with admin)
			- added functionality for URLCallback (consult with admin)
v0.1.10.4 - 08/09/2016 - fixed bug that caused browser to hang after question submission
			- fixed pdf generation of box
			- minor tweak of pdf cover page
			- added sys log on questions when they are added, edited categorised and deleted
v0.1.10.3 - 05/09/2016 - fixed pdf generation issue
			- tweaked categorisation page
v0.1.10.2 - 17/08/2016 - added left and right arrow support
			- fixed img width bug
			- fixed answer displaying incorrectly
v0.1.10.1 - 16/08/2016 - added local registration
v0.1.10 - 14/08/2016 - fixed height of images
			- redesigned reqlog requester
			- implemented moderator approval/ undo system
			- redesigned recent tab
v0.1.9.3 - 11/08/2016 - added answer keys to generated pdf
			- fixed categorise new label not bound
			- fixed subscription notification
v0.1.9.2 - 10/08/2016 - fixed some issues with generating certain style of questions
			- added notification, allowing users to see what has been changed
			- improved chat functionalities. fixed some chat bugs
			- caught a cold
v0.1.9.1 - 31/07/2016 - fixed issue with curriculum select questions
			- fixed issue of adding more blocks to view/generate
v0.1.9 - 29/07/2016 - fixed issue with choosing dp
			- added categorising feature
v0.1.8.4 - 25/07/2016 - added algorithm to prioritise questions that have not appeared (including new questions) + have not been removed often
			- fixed minor bug where view go does not generate questions
v0.1.8.3 - 21/07/2016 - redesigned login page
v0.1.8.2 - 20/07/2016 - added exhaustive method of generating questions (implement by adding "exhaustive" in notes1 of the profile page), implemented clipboard pasting of images.
v0.1.8.1 - 17/07/2016 - redesigned login panel + pdf generation. Now, questions with multiparts (a), (b), (ii) etc can be broken up. 
v0.1.8 - 16/07/2016 - added OCR support for pdf/images upload
v0.1.7 - 13/07/2016 - added pdf export, added favicon
v0.1.6.1 - 10/07/2016 - added profile page. different users with different levels of admin levels can user UI to control the functionalities of webapp
			- slightly redesigned UI. Now pages utilises EJS to render page. 
v0.1.6 - 07/07/2016 - added comment system for individual questions and general chat
			- refined user delete/edit questions
v0.1.5.1 - 03/07/2016 - added passportjs socketio. implemented possibility for moderator control
v0.1.5 - 01/07/2016 - redesigned login page. added random question functionality
v0.1.4 - 30/06/2016 - added authentication (google + facebook)
v0.1.3 - 26/06/2016 - fixed random repeats questions. Now if number of random questions requested are more than available questions, only the max number of questions will be shown.
`			- added background
v0.1.2 - 22/06/2016	- implemented new tag [spaces INTEGER lines|box|blank]
			- implemented preview/view marks
			- fixed bug where upon submitting, the more space and less space buttons are reset
			- minor fix changed the way anchor on navbar functions. so now, middle clicks should work as per normal
			- added the tooltip in view/generate > option
			- added about page
			- added guard against double submit
			- fixed potential sql injection exploit
v0.1.1 - 21/06/2016 - fixed bug:saving question without image crashes the server. bugfix: underscores in filename should now display properly
v0.1.0 - 15/06/2016 - Initial commit. 

genedSelective.js

var firstBoxX = 50;
var firstBoxY = 160;
var boxWidth = 150;
var boxPadding = 20;

var topMargin = 15;
var leftMargin = 15;
var rightMargin = 15;

var ellipseW1 = 10;
var ellipseW2 = 6;

var shortCode;

var boxLineWidth = 1;
var boxLineColor = '#ccc'

var circleLineWidth = 1;
var circleLineColor = '#ccc'

module.exports = function(socket,doc,pdfConfig,mode){
	switch(mode){
		case 'coverpage':
			socket.request.user.notes1.replace(/customShortcode\:.*?\;/,function(s){
				shortCode = s.replace(/customShortcode\:|\;/g,'')
			})
			
			socket.request.user.notes1.replace(/customCounter\:.*?\;/,function(s){
				shortCode += s.replace(/customCounter\:|\;/g,'')
			})
			
			doc.image('./public/img/gened.png',50,50,{fit : [75,75]});
			doc.font('Helvetica').fontSize(32).text('Generation Education',125,95);
			doc.font('Times-Roman').fontSize(50).text('Mock Selective Exam',60,380,{
				width : 500,
			})
			doc.font('Times-Roman').fontSize(14).text('Short Code:'+shortCode,60,660);
			doc.font('Times-Roman').fontSize(14).text('Powered by examcopedia',60,680);
			doc.font('Times-Italic').fontSize(12);
			doc.text('gen-ed.com.au',80,doc.y);
			doc.text('jimmy@gen-ed.com.au');
			
			doc.fontSize(12).font('Times-Roman');
			doc.lineWidth(pdfConfig.lineWidth);
		break;
		case 'header':
			doc.font('Times-Italic').text('Generation Education',50,38,{width : doc.width, align:'center'});
		break;
		case 'pdf info':
			doc.info['Title'] = 'mock exam';
			doc.info['Author'] = 'Generation Education, powered by Examcopedia, put together by pandamakes';			
		break;
		case 'pre end':
			doc.addPage();
			doc.image('./public/img/gened.png',50,50,{fit : [75,75]});
			//doc.font('Helvetica').fontSize(10).text('Generation Education',50,120,{width:75, align:'center'});
			doc.font('Helvetica').fontSize(32).text('Answer Sheet',125,95,{align:'left'});
			
			doc.fontSize(10).text('First name: ________________  Family name: _________________  Date: ____________  Class: ________',50,140,{align:'left'})
			
			doc.fillColor('#000').font('Helvetica').fontSize(16).text('English',firstBoxX+leftMargin,firstBoxY+topMargin,{width:boxWidth-leftMargin-rightMargin,align:'center'})
			doc.y += 12;
			for (var i = 1; i<21; i++){
				drawCircle(firstBoxX,doc.y,i,doc)
			}
			doc.strokeColor(boxLineColor).lineWidth(boxLineWidth).lineJoin('round').rect(firstBoxX,firstBoxY,boxWidth,doc.y-firstBoxY).stroke();
			
			doc.fillColor('#000').font('Helvetica').fontSize(16).text('Mathematics',firstBoxX+boxWidth+boxPadding+leftMargin,firstBoxY+topMargin,{width:boxWidth-leftMargin-rightMargin,align:'center'})
			doc.y += 12;
			for (var i = 1; i<21; i++){
				drawCircle(firstBoxX+boxWidth+boxPadding,doc.y,i,doc)
			}
			doc.strokeColor(boxLineColor).lineWidth(boxLineWidth).lineJoin('round').rect(firstBoxX+boxWidth+boxPadding,firstBoxY,boxWidth,doc.y-firstBoxY).stroke();
			
			doc.fillColor('#000').font('Helvetica').fontSize(16).text('GA',firstBoxX+2*boxWidth+2*boxPadding+leftMargin,firstBoxY+topMargin,{width:boxWidth-leftMargin-rightMargin,align:'center'})
			doc.y += 12;
			for (var i = 1; i<31; i++){
				drawCircle(firstBoxX+2*boxWidth+2*boxPadding,doc.y,i,doc)
			}
			doc.strokeColor(boxLineColor).lineWidth(boxLineWidth).lineJoin('round').rect(firstBoxX+2*boxWidth+2*boxPadding,firstBoxY,boxWidth,doc.y-firstBoxY).stroke();
			
		break;
	}
	return shortCode;
}

function drawCircle(left,top,idx,doc){
	doc.fillColor('#666').fontSize(12).text(idx,left+leftMargin,top,{width:20,align:'center'})
	
	for (var j = 0; j<4; j++){
		doc.fillColor('#666').fontSize(8).text(mc[j],left+35+j*25,top+1,{width:ellipseW1*2,align:'center'});
		doc.strokeColor(circleLineColor).lineWidth(circleLineWidth).ellipse(left+35+j*25+ellipseW1,top+4,ellipseW1,ellipseW2).stroke()
	}
	
	if((idx)%5==0){
		doc.y += 12;
	}else{
		doc.y += 5;
	}
}

/* make 4 random characters */
/* http://stackoverflow.com/a/1349426/6059235 */
function makeid(){
    var text = "";
    var possible = "ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz23456789";

    for( var i=0; i < 4; i++ )
        text += possible.charAt(Math.floor(Math.random() * possible.length));

    return text;
}

var mc = ['A','B','C','D'];